packages/signers/src/capsule/signer.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  WalletClientSigner,
+  type SignTypedDataParams,
+  type SmartAccountAuthenticator,
+} from "@alchemy/aa-core";
+import Capsule, { createCapsuleViemClient } from "@usecapsule/web-sdk";
+import { createWalletClient, custom, type Hash, type WalletClient } from "viem";
+import type {
+  CapsuleAuthenticationParams,
+  CapsuleConfig,
+  CapsuleUserInfo,
+} from "./types.js";
+
+/**
+ * This class requires the `@usecapsule/web-sdk` dependency.
+ * `@alchemy/aa-signers` lists it as an optional dependency.
+ *
+ * @see: https://docs.usecapsule.com/getting-started/initial-setup
+ */
+export class CapsuleSigner
+  implements
+    SmartAccountAuthenticator<
+      CapsuleAuthenticationParams,
+      CapsuleUserInfo,
+      Capsule
+    >
+{
+  inner: Capsule;
+  private client: WalletClient;
+  private signer: WalletClientSigner | undefined;
+
+  constructor(
+    params: CapsuleConfig | { inner: Capsule; client: WalletClient }
+  ) {
+    if ("inner" in params) {
+      this.inner = params.inner;
+      this.client = params.client;
+      return;
+    }
+
+    this.inner = new Capsule(params.env, params.apiKey, params.opts);
+    this.client = createCapsuleViemClient(
+      this.inner,
+      params.walletConfig as any, // TODO: Capsule team to address lint error
+      params.viemClientOpts
+    ) as unknown as WalletClient; // TODO: Capsule team to address lint error
+  }
+
+  readonly signerType = "capsule";
+
+  getAddress = async () => {
+    if (!this.signer) throw new Error("Not authenticated");
+
+    const address = await this.signer.getAddress();
+    if (address == null) throw new Error("No address found");
+
+    return address as Hash;
+  };
+
+  signMessage = async (msg: Uint8Array | string) => {
+    if (!this.signer) throw new Error("Not authenticated");
+
+    return this.signer.signMessage(msg);
+  };
+
+  signTypedData = (params: SignTypedDataParams) => {
+    if (!this.signer) throw new Error("Not authenticated");
+
+    return this.signer.signTypedData(params);
+  };
+
+  authenticate = async () => {
```

Comment:
I couldn't find where a Capsule user is either created or logged into an existing Capsule account. Should this method be handling initiating a Capsule user session? If so we would need to use a user's email to log them in or create a new account by using the methods in our documentation